Vue.jsを使ったTodoリストの作成

Masaya-Nishijima opened the pull request.

Masaya-Nishijima committed:
feat: HTML, CSS のみで、デザインの作成

Masaya-Nishijima committed:
feat: JavaScript(Vue.js)のセットアップ

Masaya-Nishijima committed:
feat: Veu.jsを使ってメモリストを表示

Masaya-Nishijima committed:
feat: メモの削除機能を追加

Masaya-Nishijima committed:
refactor: 不要なコメントを削除

Masaya-Nishijima committed:
fix: 常に一番上のメモから削除されるバグを修正

バグの要因: メソッドに変数を渡していない

Masaya-Nishijima committed:
refactor: スペルミスを修正

Masaya-Nishijima committed:
feat: メモの作成機能を追加

Masaya-Nishijima committed:
feat: 編集機能の作成

Masaya-Nishijima committed:
refactor: ToDo作成の入力欄がnew todoに固定化

これまでは、直近に作成したTodoと同じになっていた。

Masaya-Nishijima committed:
feat: inport, exportメソッド作成(HTMLへは未反映)

Masaya-Nishijima committed:
feat: Todoの保存、読み出し機能完成

Masaya-Nishijima committed:
style: HTMLを整形

Masaya-Nishijima committed:
fix: 無効なCSSを削除

Masaya-Nishijima committed:
refactor: CSSの改善

Masaya-Nishijima committed:
fix&refactor: importの動作不良修正&app名変更

Masaya-Nishijima committed:
refactor: v-forで :keyを使用する

Masaya-Nishijima committed:
refactor: 複数の属性を持つHTML要素は、属性毎に改行する。(スタイルガイドより)

Masaya-Nishijima committed:
refactor: インポートエクスポートのtodoをtodosへ変更

Masaya-Nishijima commented:
v-for v-if

in ex でいちぶだけ、ならびずれる(3over)

Masaya-Nishijima committed:
style: Vue.js スクリプトの読み込みの記述位置移動

defer属性を付与することでDOMツリー生成後にスクリプトの実行を指示できたので、
修正

Masaya-Nishijima committed:
refactor: JavaScript Standard Style

Masaya-Nishijima committed:
算出プロパティの利用

Masaya-Nishijima commented on index.html:
```diff
@@ -0,0 +1,56 @@
+      </div>
+      <div class="body">
+        <ul class="todo-list">
+          <li
+            class="todo-list-item"
+            v-for="(todo, index) in todos"
+            :key="todo.id"
+          >
+            <div class="todo-list-item-body"></div>
+            <em v-if="!todo.edit">
+              {{ todo.text }}
+            </em>
+            <input
+              type="text"
+              v-else
+              v-model="todos[index].text"
+            >
```
同一要素で`v-if`, `v-for`を同時に使用しないと、スタイルガイドにありましたが、
こちら親子関係の要素で、使用してしまっています。
この場合も禁止される同時使用にあたるのでしょうか?

maedana replied:
liでv-for、その入れ子のemでv-ifですし、同一要素でとはならないでかなと思いました。
これですよね?
https://v3.ja.vuejs.org/style-guide/#v-for-と一緒に-v-if-を使うのを避ける-必須

Masaya-Nishijima replied:
はい、そちらのリンクの箇所の禁止事項に該当しないか疑問でした。

回答ありがとうございます。

Masaya-Nishijima commented on script.js:
```diff
@@ -0,0 +1,40 @@
+    changeMode: function (todoIndex) {
+      this.todos[todoIndex].edit = !this.todos[todoIndex].edit
+    },
+    importTodos: function () {
+      this.todos = []
+      for (let index = 0; index < localStorage.length; index++) {
+        const todo = {
+          text: localStorage.getItem(localStorage.key(index)),
+          edit: false
+        }
+        this.todos.push(todo)
+      }
```
3つ以上のTodoを、インポートしようとすると一部のTodoの順番が入れ替わってしまいます。(原因不明)

一部のTodo --> 1つ目のTodoは問題ないが、2つ目以降のTodoが一定のパターンで入れ替わっているようです。

maedana replied:
入れ替わるとは、`this.todos.push(todo)`した後の`this.todos`の並びが、`localStorage`に保存されている順と変わってしまうってことですか? 

https://developer.mozilla.org/ja/docs/Web/API/Storage/key
によると
> キーの順序はユーザエージェント依存であり、この順序に頼るべきではありません。

とあるのが気になりますね。あとそもそもで言うと、localStorageってドメインごとの管理なので、同一ドメインで他用途でも利用する可能性を考慮すると、全部のキーを列挙する実装自体あまり良くないんじゃないかなって思いました。いかがでしょうか?

Masaya-Nishijima replied:
単一キー中にすべてのデータが格納されるように修正したところ、無事順番が入れ変わる不具合が解消できました。
![スクリーンショット 2022-06-15 14 23 34](https://user-images.githubusercontent.com/100653079/173743984-808add50-d7da-46ed-baad-a9efe51030ac.png)

元の実装にしていた理由は、以下の通りです。
- メモアプリと同一ドメインに他アプリが来ることを想定していませんでした。
  - メモの数で表示が縦に伸びるので、同一ページにアプリを並べにくい。
  - 仮にも"アプリ"なら/で区切らずにサブドメイン切るだろう。
- RDBの設計の学習をしていたため、同じカラムに複数のデータが入るのが、気持ち悪く感じていた。

Masaya-Nishijima replied:
どうしようもない問題ですが、同一カラムに文字列として保存している関係で、区切り文字(今回は`,`)がToDoに含まれていると、インポートしたタイミングで複数のメモに分裂してしまいます。

maedana replied:
データ保存形式はJSON形式にしたりすると扱いやすいですよー。
> インポートしたタイミングで複数のメモに分裂してしまいます。

という問題も起きなくなるかなーと。いかがでしょうか?

Masaya-Nishijima replied:
JSON形式にすることで、メモが分裂する問題を解決できました。
提案、ありがとうございます。
a0d5d59

Masaya-Nishijima commented on script.js:
```diff
@@ -0,0 +1,40 @@
+      })
+    }
+  },
+  methods: {
```
算出プロパティにすると以下のような問題が発生したため、データをlocalStorageに書き出す処理以外は、メソッドとして定義しています。
- 削除が一回しかできない
- 編集ができない
- 同一の内容で、連続登録できない

なにか対策はありますでしょうか?
(コンポーネントや子インスタンスの作成、データの形式を変える, etc...)

maedana replied:
現在の実装ではそもそもcomputedにするのが適切なものはなさそうですよ。全部イベントハンドラですし。

Masaya-Nishijima replied:
>現在の実装ではそもそもcomputedにするのが適切なものはなさそうですよ。

かしこまりました。課題完了後に、メンターさん方がどのように実装してcomputedを利用しているか確認します!

Masaya-Nishijima commented on reset.css:
```diff
@@ -0,0 +1,108 @@
+/*
```
CSS上級の課題のreste.cssをコピーして利用しています。(=パクり)

`public domain`とされていますが、記述されているURLの内容と異なるため確認です。

Masaya-Nishijima replied:
追記：私が書くコードにもこのようライセンス的なものを書いてたほうが良いのでしょうか?

```css
/* http://meyerweb.com/eric/tools/css/reset/
   v2.0 | 20110126
   License: none (public domain)
*/
```

maedana replied:
https://ja.wikipedia.org/wiki/パブリックドメイン
を読む限り問題なさそうですねー

Masaya-Nishijima replied:
確認ありがとうございます。

Masaya-Nishijima reviewed (commented):
プログラミングをしていて気になったことをコメントしています。
確認よろしくおねがいします。

maedana commented on index.html:
```diff
@@ -0,0 +1,56 @@
+    <title>ToDo</title>
+    <link rel="stylesheet" href="reset.css">
+    <link rel="stylesheet" href="style.css">
+    <script src="https://cdn.jsdelivr.net/npm/vue@2/dist/vue.js"></script><!-- 開発バージョンのVue.jsを使用 -->
```
vue2にした理由ってあるんでしょうか? 理由がなければvue3でいいんじゃないかなーって気もします。(プラクティス後半のチーム開発ではvue2ですけど)

Masaya-Nishijima replied:
プラクティスの資料へのリンクがVue2の物だったのでVue2を利用しています。
>[はじめに — Vue.js](https://jp.vuejs.org/)

Vue3への移行は、`script.js`を大幅に書き換える必要がありそうです。(インスタンスの定義方法が違う?)

移行作業のための差分が大量発生してしまうので、とりあえず他のレビュー箇所を、現状のVue2で修正致しますので、そのあとで、移行をすべきか意見をください。

maedana replied:
OKです。ではvue2のままでいきましょう。ただ、中長期的にはvue3を使う現場が多くなると思いますし、vue3の書き方にも慣れていくとよいかと思います。

Masaya-Nishijima replied:
承知致しました。

maedana commented on index.html:
```diff
@@ -0,0 +1,56 @@
+    <script type="text/javascript" src="script.js" defer></script>
+  </head>
+  <body>
+    <div class="main" id="todoItem">
```
htmlのidとかclass名って通常はケバブケースにするイメージがあります。少ししたの`site-title`なんかはそうなってますよね。

vueにバインドするためのidですが、todo全体を管理するものですよね。todoItemだと1件だけみたいに見えませんか? というわけで命名も見直してみて欲しいかなと思いました。

Masaya-Nishijima replied:
始め、`todos`としていたのですが、
[スタイルガイド — Vue\.js 複数単語コンポーネント名](https://jp.vuejs.org/v2/style-guide/index.html#%E8%A4%87%E6%95%B0%E5%8D%98%E8%AA%9E%E3%82%B3%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%8D%E3%83%B3%E3%83%88%E5%90%8D-%E5%BF%85%E9%A0%88)
こちらを読んでなにを思ったか`todoItem`としていました。

`todo-app`に変更します。

maedana commented on index.html:
```diff
@@ -0,0 +1,56 @@
+    <div class="main" id="todoItem">
+      <div class="header">
+        <h1 class="site-title">ToDo</h1>
+        <h2><button v-on:click="importTodos()">import</button></h2>
```
一貫性があればどっちでもいいと思いますが、個人的には短いので`@click`という書き方が好きですね。あと括弧も省略可能なので省略しますね。ご参考までに。

Masaya-Nishijima replied:
`@click="importTodos()"`
に修正致しました。他の`v-on`も同様に変更

Masaya-Nishijima replied:
`importTodos()`の括弧は、省略できないようです。括弧を無くすと下のようなエラーが出て、Vueがフリーズしてしまいます。(すべてのボタンが反応しない)
![スクリーンショット 2022-06-15 10 44 25](https://user-images.githubusercontent.com/100653079/173718489-60378256-a839-41a7-9566-2f9e21a50ff6.png)

maedana replied:
失礼しました。省略出来るのは算出プロパティでしたね :pray:

maedana commented on index.html:
```diff
@@ -0,0 +1,56 @@
+          <li class="new-todo-list-item">
+            <input
+              type="text"
+              v-model="new_todo"
```
jsの変数名はキャメルケースにしたいですねー。他はそうなってると思いますし統一しましょうー。

Masaya-Nishijima replied:
`new_todo`→`newTodo`に変更致しました。
HTML, CSS JavaScript, Vue.jsの4種類の命名規則を使い分けなければならない状況で、混乱してRubyの命名規則を使っていました。🙇

maedana replied:
> 4種類の命名規則を使い分けなければならない状況

ですよねー...。とはいえ長年の文化なので乗っかっておくのが無難かなーと思ってなるべく乗っかるようにしております。

maedana commented on style.css:
```diff
@@ -0,0 +1,44 @@
+* {
+  border-radius: 15px;
+}
```
全要素にこれをつけるのはいくら利用している要素が少ないとはいえ乱暴すぎないですかね...? 長期的にメンテナンスするソフトウェアだとすると早々に破綻しそうですし、課題といえどやめたほうがいいんじゃないかなって思いました。いかがでしょうか?

Masaya-Nishijima replied:
さすがにDRYすぎますね。(DRYの意味をはき違えるている。)

下のように`todo-app`そのものと、子孫要素の`button`, `input`にのみ掛るように修正致しました。
```CSS
#todo-app, #todo-app button, #todo-app input{
  border-radius: 15px;
}
```

maedana commented on script.js:
```diff
@@ -0,0 +1,40 @@
+  data: {
+    todos: [
+      { text: 'memo 1', edit: true },
+      { text: 'memo 2', edit: false }
```
editで表示制御してますが、htmlの見た目上の都合であって、永続化するデータに含めるのはちょっと違うかなーと思いました。いかがでしょうか?

Masaya-Nishijima replied:
> editで表示制御してますが、htmlの見た目上の都合であって、永続化するデータに含めるのはちょっと違うかなー

この考え方には同意です。また、実装に当っても、localStorageにはeditの情報は乗らないよう実装しています。
```vue.js
this.todos.forEach((todo, index) => {
  localStorage.setItem(index, todo.text) //本文(.text)のみ保存してeditは残らない。
})
```
ただし、`永続化するデータ=localStorage`と考え、インスタンスの`data:`はあくまで一時的なデータと認識していました。そのため、このような実装になっています。

Vue.jsのインスタンス下のデータも永続化するデータと考えるべきでしょうか?
また、その場合はどこにeditのデータを保持すべきでしょうか?

maedana replied:
> Vue.jsのインスタンス下のデータも永続化するデータと考えるべきでしょうか?
また、その場合はどこにeditのデータを保持すべきでしょうか?

localStorageに入らなければいいと思います。

あとは例えば`editingIndex`のような命名で編集中のindex1件だけ持つのはどうでしょう? 同時編集が一件だけという制限になりますし、仕様変わっちゃいますけど。ただ、todosの内部データとして持つ必要ななくなるかな、と。

Masaya-Nishijima replied:
こちらの`editingTodo`で編集中のメモを指定するように修正しました。


```javascript
  data: {
    todos: [],
    editingTodo: {index: null},
```
15665c4
![スクリーンショット 2022-06-17 10 57 47](https://user-images.githubusercontent.com/100653079/174207204-09a83d1c-3ab6-4778-a8d3-24a93e3f7595.png)
![スクリーンショット 2022-06-17 10 58 58](https://user-images.githubusercontent.com/100653079/174207330-22e71390-ddb7-465b-89cb-acb601289861.png)

##
こちらの制限につきましては、
>同時編集が一件だけという制限

逆に、[こちらの修正案](https://github.com/Masaya-Nishijima/study_Vue.js/pull/1#discussion_r898807183)を、自動的に満せるのでメリットかもしれないです。

maedana commented on script.js:
```diff
@@ -0,0 +1,40 @@
+      { text: 'memo 1', edit: true },
+      { text: 'memo 2', edit: false }
+    ],
+    new_todo: 'new todo'
```
上述した通りですがjsですし、キャメルケースにしましょう。

Masaya-Nishijima replied:
[こちら](https://github.com/Masaya-Nishijima/study_Vue.js/pull/1#discussion_r897456679)のように修正しています

maedana commented on script.js:
```diff
@@ -0,0 +1,40 @@
+      this.todos.forEach((todo, index) => {
+        localStorage.setItem(index, todo.text)
+      })
+    }
```
算出プロパティにするのは基本は`data`から導出出来るリアクティブにしたいもの、かなって思います。各種イベントの処理っていうのは別にリアクティブにしたいものでもなんでもないのでmethodsに書きましょう。

そもそもで言うと、exportTodos, importTodosでlocalStorageから読み書きとなってますが、そうではなくて、todoを作成/更新/削除するたびにlocalStorageを自動更新する仕組みにしてほしいかなと思います。作成だけしてブラウザ閉じたらデータが失われるとか悲しくないですかね?

maedana replied:
メソッド定義は`exportTodos()`のように書くことも可能です。個人的には短いのでこちらが好みですね。ご参考までにー

Masaya-Nishijima replied:
- methodsに書きましょう
  - [x] expartTodosを算出プロパティから、メソッドに移動させました。
- todoを作成/更新/削除するたびにlocalStorageを自動更新する仕組みにしてほしいかなと思います。
  - [x] Create Update Delete のメソッドの最後にexportTodosを実行するように変更致しました。
  - 各メソッドの最終行に書くより`updated:` か `beforeUpdate:`どちらかに記述したほうがDryになりますかね?
- `exportTodos()` のように書くことも可能です
  - [x] コメントありがとうございますこちらの形式に変更させて頂きました。
  JavaScript Vue.jsで、どんな構文が好んで使われているのか理解っていなかったので情報ありがたいです!

Masaya-Nishijima replied:
> 作成だけしてブラウザ閉じたらデータが失われるとか悲しくないですかね?

たしかに、WordやExcelの感覚で考えてと自動保存すべきでした。

とりあえずで、localStorageを利用できるコードを作成しようとして、ボタンで保存・呼び出しをするように実装しました。そして、「毎回固定のタスクを呼びだして終った項目を削除する」使い方を想定すれば、元のコードでもいいや!
(ex. 毎日朝一番にimport{朝御飯, 昼御飯, 夜御飯, お風呂}して、各作業が終ったら削除していく のような)

っと投げ遣りになってしまいリファクタリングが足りていませんでした。

maedana replied:
> 各メソッドの最終行に書くよりupdated: か beforeUpdate:どちらかに記述したほうがDryになりますかね?

各メソッドの最終行に書くのがわかりやすくていいと思います :smile:

maedana commented on index.html:
```diff
@@ -0,0 +1,56 @@
+            >
+            <button v-on:click="changeMode(index)">
+              編集
+            </button>
```
編集押すたびに表示とフォームと切り替わるようですが、直感的にわからないかなーと感じました。フォームの状態を終わるのが「編集」ボタンというのはわかりにくくないですかね?

Masaya-Nishijima replied:
![スクリーンショット 2022-06-15 10 47 47](https://user-images.githubusercontent.com/100653079/173719109-5fd3d55d-47d4-4fe7-ab05-67eb437ef083.png)
こちらでどうでしょうか?
`編集完了ボタン`や、`保存ボタン`を作成することも考えましたが、
実装がフォームに入力された段階でインスタンスに記録されているため、
保存ボタンを押す前なのに保存されているという矛盾が発生するので
このような形にしてみましたがいかがでしょうか?

追記 `表示ボタン`単体でない理由
編集中に、`表示ボタン`単体があっても、なにを表示するんだよ?
ってなりそうなの`編集と表示`の切り替えであることが、わかるように意識してこの形にしています。

maedana replied:
> 実装がフォームに入力された段階でインスタンスに記録されているため、
保存ボタンを押す前なのに保存されている

こうならないようにしてみて欲しいですが、難しいでしょうか?

Masaya-Nishijima replied:
編集中に保存される要因は、他の要素の変更に合せて保存されてしまうことでした。
ex. 
1. `Todo A`を編集中に、`Todo B`を編集開始or編集終了, 保存する。
2. `Todo A`を編集中に、`新規のTodo`を作成する。
こちらは、あまり発生しないため、今回修正対象として考慮に入れてないです。

[こちらの修正で](https://github.com/Masaya-Nishijima/study_Vue.js/pull/1#discussion_r898799711)メモを一つずつしか編集できなくした結果、問題が発生しなくなったので自動的に解決しました。

Masaya-Nishijima committed:
refactor: クラス名をケバブケースに

Masaya-Nishijima committed:
style: `v-on:click`を`@click`に変更

Masaya-Nishijima committed:
refactor: 編集フォームの終了が編集ボタンではなく編集←→表示ボタンに変更

Masaya-Nishijima committed:
refactor: JSの変数名をキャメルケースに

Masaya-Nishijima committed:
refactor: import exportを自動化

Masaya-Nishijima committed:
style: CSSで要素の角丸めの適応範囲を狭める。

Masaya-Nishijima committed:
refactor: 不要になっていたimportコマンドを削除

Masaya-Nishijima committed:
refactor: localStorageでは単一カラムにデータが入るように修正

Masaya-Nishijima reviewed (commented):
[Vueのバージョンの件](https://github.com/Masaya-Nishijima/study_Vue.js/pull/1#discussion_r896666862)以外、
頂いたレビューに沿って修正しています。

Masaya-Nishijima committed:
fix: 編集機能の処理を大幅に変更

- todosに編集中か否かのフラグを保持しないように
- 編集中のファイルがある場合に削除ボタンが出ないように
各todoが編集フラグを持たないようになったため、削除動作と編集動作を同時に行うと表示がズレるため。

Masaya-Nishijima committed:
fix: JSON形式を使い、ローカルストレージに保存することで使えない文字(区切り文字)をなくす
